Wikipedia/Code/SavedArticlesFetcher.m

```diff
@@ -163,18 +179,91 @@ - (void)fetchTitle:(MWKTitle*)title failure:(WMFErrorHandler)failure success:(WM
 }
 
 - (void)downloadImageDataForArticle:(MWKArticle*)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
+    if (![[NSUserDefaults standardUserDefaults] wmf_didFinishLegacySavedArticleImageMigration]) {
+        WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.0.5)
+        [self migrateLegacyImagesInArticle:article];
+    }
     [self fetchAllImagesInArticle:article failure:^(NSError *error) {
         failure([NSError wmf_savedPageImageDownloadError]);
     } success:^{
         [self fetchGalleryDataForArticle:article failure:failure success:success];
     }];
 }
 
+- (void)migrateLegacyImagesInArticle:(MWKArticle *)article {
+    //  Removes up old saved article image list folders, copies old cached images to original and article image width locations. This ensures articles saved with 5.0.4 and older will still have images availble offline in 5.0.5. The migration is idempotent - the enumerated folders are removed so they won't be processed the next time around.
+    
+    //Get the folder that contains legacy saved article images for example - articles/Barack_Obama/Images/
+    NSString *imagesFolderPath = [self.savedPageList.dataStore pathForImagesWithTitle:article.title];
+    NSURL *imagesFolderURL = [NSURL fileURLWithPath:imagesFolderPath isDirectory:YES];
+    NSEnumerator *enumerator = [[NSFileManager defaultManager] enumeratorAtURL:imagesFolderURL includingPropertiesForKeys:@[NSURLIsDirectoryKey] options:NSDirectoryEnumerationSkipsSubdirectoryDescendants errorHandler:^BOOL(NSURL * _Nonnull url, NSError * _Nonnull error) {
+        DDLogError(@"Error enumerating image directory: %@", error);
+        return YES;
+    }];
+    
+    WMFImageController *imageController = [WMFImageController sharedInstance];
+    NSUInteger articleImageWidth = [[UIScreen mainScreen] wmf_articleImageWidthForScale];
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    
+    for (NSURL *imageFolderURL in enumerator) { //Enumerate each subfolder of the image folder. There is one subfolder per saved image.
```

Comment:
I think it's still ok to blast those folders, though, to keep it idempotent. 

However, we probably don't want to blast this Image**s**.plist file until the article record is being deleted. This is because it is the only record we have of these 5.0.4 style image urls. The reason we still need to know about them is so we can clear the SDWebImage cache of these images when we're deleting article data. 

For example, I did a clean install on 5.0.4, loaded and saved obama article, stopped the app, updated to this PR, started app, the **un**saved obama and cleared history, then stopped the app. It appeared that the obama images were removed from SDWebImage's cache folder *except* for the old 5.0.4 images - but the Image**s**.plist file has those urls, so I think we can just make sure we try to also delete any images from that plist at the time we're clearing out other article image data and we should be good. 

I hope that makes sense...